### PR TITLE
Storing certs to chainstate_db

### PR DESCRIPTION
**Pull request by abi87**
Extending `CCoins `class to include sidechains related information is going to be needed for efficient operations on sidechains history.
This branch is a current proof of concept to validate the idea above.
Requesting @alsala an intermediate review, while code still needs to cleanup.
Full regression successfully performed.

### CODE REVIEW

File: src/coins.h [hunk 1]
```diff
@@ -31,14 +31,20 @@ class CBlockUndo;
  * - unspentness bitvector, for vout[2] and further; least significant byte first
```
Comment (alsala) [inline src/coins.h:31]: formatting of comments: '-' is used as paragraph marker, remove it if the phrase contonues

File: src/coins.cpp [hunk 2]
```diff
@@ -14,11 +14,107 @@
...
+bool operator!=(const CCoins &a, const CCoins &b) {
+    return !(a == b);
+}
+
+bool CCoins::IsCoinBase() const {
+    return fCoinBase;
+}
+
+bool CCoins::IsFromCert() const {
+    // when restored from serialization, nVersion is populated only with latest 7 bits of the original value!
```
Comment (alsala) [inline src/coins.cpp]: this is true only with negative versions, pls update the comment

File: src/coins.h [hunk 3]
```diff
@@ -244,33 +250,24 @@ class CCoins
...
+            ::Unserialize(s, originScId,       nType,nVersion);
+            ::Unserialize(s, bwtOutputCounter, nType,nVersion);
+        }
 
-    //! check whether a particular output is still available
-    bool IsAvailable(unsigned int nPos) const {
-        return (nPos < vout.size() && !vout[nPos].IsNull());
-    }
+        for(unsigned int idx = 0; idx < vout.size(); ++idx)
+            if ( idx < bwtOutputCounter)
```
Comment (alsala) [inline src/coins.h]: Pls check that the order of out type (from cert or not) in vout is the same as in CScCertificate class

File: src/txdb.cpp [hunk 4]
```diff
@@ -279,6 +279,10 @@ bool CCoinsViewDB::GetStats(CCoinsStats &stats) const {
                         nTotalAmount += out.nValue;
                     }
                 }
+
```
Comment (alsala) [inline src/txdb.cpp:282]: Add coins.bwtOutputCounter member

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -1835,21 +1835,14 @@ CAmount CWallet::GetDebit(const CTxIn &txin, const isminefilter& filter) const
 {
     {
         LOCK(cs_wallet);
-#if 0
```
Comment (alsala) [inline src/wallet/wallet.cpp:1838]: Consider restoring previous implementation that might be slightly more efficient even if it was less readable

File: src/wallet/wallet.h [hunk 6]
```diff
@@ -474,7 +474,7 @@ class CWalletObjBase : virtual public MerkleAbstractBase
     virtual void SetNoteData(mapNoteData_t &noteData) {}; // default is null
 
     //! filter decides which addresses will count towards the debit
-    virtual CAmount GetDebit(const isminefilter& filter) const { return 0; } // certs has no vin
+    virtual CAmount GetDebit(const isminefilter& filter) const;
```
Comment (alsala) [inline src/wallet/wallet.h]: GetDebit could now be a non-virtual base class method

**Review by alsala — APPROVED** [review]

File: src/wallet/wallet.h (hunk 6 above)
Comment (abi87) [reply src/wallet/wallet.h]: Done

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (abi87) [reply src/wallet/wallet.cpp:1838]: Agreed & done

File: src/txdb.cpp (hunk 4 above)
Comment (abi87) [reply src/txdb.cpp:282]: Done

File: src/coins.h (hunk 3 above)
Comment (abi87) [reply src/coins.h]: Done

File: src/coins.cpp (hunk 2 above)
Comment (abi87) [reply src/coins.cpp]: Done

File: src/coins.h (hunk 1 above)
Comment (abi87) [reply src/coins.h:31]: Done

**Pull request merged**